feat: Sidebar Attachments accessibility

marination opened the pull request:
`no-docs`

> docs will be added after merge. wiki was very buggy when i tried

###  Issue: 

- With a large volume of attachments, the "Attach File" button is pushed to the bottom
- It's hard to search through these files from the Form
  ![2023-07-26 20 13 24](https://github.com/frappe/frappe/assets/25857446/f77f96a1-1a91-408b-b91c-2fb7e5a68cc8)


### Enhancement
- "Attach File" stays at the top of the pile. Last 10 (latest) attachments are shown with "Show All" button
   ![2023-08-09 18 35 44](https://github.com/frappe/frappe/assets/25857446/b6fee741-4576-43b6-87d0-8f4ba50f41cf)

  

- Small explore files button added so that users can use the File View to navigate/filter through files
  ![2023-07-26 20 02 43](https://github.com/frappe/frappe/assets/25857446/ef8a7fe3-07e0-49b8-bd8c-950397c9bb4b)

- Expanded Explore Files button when the max limit is reached
  ![2023-07-26 20 04 42](https://github.com/frappe/frappe/assets/25857446/095667eb-90a8-416d-bf1c-79f4180725f0)

- Added `file_type` to Files, this is useful for filtering and visibility
   <img width="1321" alt="Screenshot 2023-07-26 at 8 06 05 PM" src="https://github.com/frappe/frappe/assets/25857446/f07618ab-4a67-4200-8d73-6ce75c2cb4ef">
- Patch to set File Type  in all files

marination committed:
feat: Sidebar Attachments accessibility

- Issue: With a large volume of attachments, the "Attach File" button is pushed to the bottom
- "Attach File" stays at the top of the pile
- Small explore files button added so that users can use the File View to navigate/filter through files
- Expanded Explore Files button when attach file action is hidden
- Added `file_type` to Files, this is useful for filtering and visibility
- Added "Type" to File List View
- Patch to set File Type  in all files

ankush commented on frappe/patches/v15_0/set_file_type.py:
```diff
@@ -0,0 +1,18 @@
+		file_name = file.get("file_name").split("?")[0]
+		file_extension = file_name.split(".")[-1].upper() if file_name.split(".")[-1] else None
+		if file_extension:
+			frappe.db.set_value("File", file.get("name"), "file_type", file_extension)
```
```suggestion
			frappe.db.set_value("File", file.get("name"), "file_type", file_extension, update_modified=False)
```

Also, probably good idea to enable auto-commit since this can cause arbitrarily high number of queries.

ankush commented on frappe/patches/v15_0/set_file_type.py:
```diff
@@ -0,0 +1,18 @@
+			continue
+
+		file_name = file.get("file_name").split("?")[0]
+		file_extension = file_name.split(".")[-1].upper() if file_name.split(".")[-1] else None
```
mimetypes.guess_type and guess_extension (?)

marination committed:
fix: Use `mimetypes` to set file type & auto commit in patch

Co-authored-by: Ankush Menat <ankush@frappe.io>

marination committed:
feat: Add "Show All" button and render limited attachments on refresh

- Render last 10 (latest) attachments and show "Show All" button
- "Show All" will re-render and show all attachments
- This state of showing all attachments will be maintained until next page refresh

marination commented:
Failing awesome bar test is unrelated. Also seen in https://github.com/frappe/frappe/actions/runs/5706335436/job/15461930152?pr=21846

marination committed:
test: Sidebar attachments UX

- Added 10 royalty free images from 'https://unsplash.com/images/stock/royalty-free'
- Test to check state of attachments actions through different user actions
- UI Helper function to create ToDo and add max attachment limit of 12

surajshetty3416 committed:
Merge branch 'develop' into attachments-access

barredterra commented on frappe/patches/v15_0/set_file_type.py:
```diff
@@ -0,0 +1,34 @@
+
+def execute():
+	"""Set 'File Type' for all files based on file extension."""
+	files = frappe.db.get_all(
+		"File",
+		fields=["name", "file_name", "is_folder"],
+	)
+
+	frappe.db.auto_commit_on_many_writes = 1
+
+	for file in files:
+		if file.get("is_folder"):
+			continue
+
+		file_extension = get_file_extension(file.get("file_name"))
+		if file_extension:
+			frappe.db.set_value(
+				"File", file.get("name"), "file_type", file_extension, update_modified=False
+			)
```
```suggestion
	files = frappe.db.get_all(
		"File",
		fields=["name", "file_name", "file_url"],
		filters={
			"is_folder": 0,
			"file_type": ("is", "not set")
		},
	)

	frappe.db.auto_commit_on_many_writes = 1

	for file in files:
		file_extension = get_file_extension(file.file_name or file.file_url)
		if file_extension:
			frappe.db.set_value(
				"File", file.name, "file_type", file_extension, update_modified=False
			)
```

- Some files have a URL but no file name
- Get only files that are not folders and are missing the file type
- `get_all` returns a list of `frappe._dict`s, so we can access values via `.` instead of `get()`

marination replied:
> Some files have a URL but no file name

Using `file_url` directly to set the file type as the `file_name` is extracted from the last part of the URL.

barredterra reviewed (commented):
Sample attachments don't need to be big and pretty, since only the computer will "see" them during automated tests. 😄 Single pixels and/or plain text files with a couple of characters should be enough.

barredterra commented on frappe/public/js/frappe/form/templates/form_sidebar.html:
```diff
@@ -54,17 +54,45 @@
+		</li>
+
+
+		<li class="show-all-btn hidden">
+			<button class="data-pill btn">
+				<span class="pill-label ellipsis">
+					{%= __("Show All") %}
+				</span>
+			</button>
+		</li>
```
```suggestion
		<li class="show-all-btn hidden">
			<a href="" class="pill-label ellipsis">
				{%= __("Show All") %}
			</a>
		</li>
```

This button feels a bit too prominent/distracting to me. What do you think about formatting it as a link?

![Bildschirmfoto 2023-08-04 um 16 40 55](https://github.com/frappe/frappe/assets/14891507/0ce1a453-0547-40cf-8b36-11d946492070)

barredterra commented:
For files like this (no file extension, but a valid URL), should we set the file type to "URL"?

![Bildschirmfoto 2023-08-04 um 16 44 06](https://github.com/frappe/frappe/assets/14891507/3c506d24-eba3-40d9-846e-bafb3a59ce4f)

sagarvora commented:
> For files like this (no file extension, but a valid URL), should we set the file type to "URL"?

[No](https://file.org/extension/url)

marination committed:
fix: Make Show All button less prominent

marination committed:
fix: Use File URL to set file type as File Name is not reliable

marination commented:
> For files like this (no file extension, but a valid URL), should we set the file type to "URL"?

Link is better (its not a valid file extension afaik) ?

marination committed:
chore: Use lighter files as test fixtures

sagarvora commented:
> Link is better (its not a valid file extension afaik) ?

I thought more about this. I think a user would intuitively expect a file extension even for a remote URL and it'd be inconsistent to show the type as Link/URL only for some of the links. How about a separate read-only checkbox for links instead?

(this could be a separate PR with updates to File controller as well)

marination committed:
fix: Correct value for `guess_type` and empty sample attachments
